---
title: "Integrating human and single-cell immune variations reveals unadjuvanted vaccine high responders are naturally adjuvanted"
output: 
  html_document:
    code_folding: hide
    theme: flatly
    highlight: kate
  github_document:
    html_preview: true
    toc: TRUE
---

```{r, include = FALSE}
library(here)
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE, 
  eval = FALSE,
  root.dir = here()
)
```


**Code to reproduce all manuscript results and figures**   
Matt Mul√®

## Table of Contents
1.  [Instructions for analysis workflow](#instructions)  
2.  [Fig1 & Fig S1. sample, cell frequency, and protein distributions](#fig1.1)  
3.  [Fig S1. transcriptome analysis of manually gated plasmablasts and activated B cells](#fig1.2)  
4.  [Fig 1. multivariate analysis of human and cell type variations](#fig1.3)  
5.  [Fig 2 & Fig S2. mixed effects timed vaccination response model -- unadjuvanted cohort](#fig2.1)  
6.  [Fig S2 visualization of day 7 post vaccination phenotypes and predictive signature deconvolution](#fig2.2)
7.  [Fig 2. bottom up single cell reconstruction of single cell monocyte pseudotime](#fig2.3)  
8.  [Fig 3. & FigS3 mixed effects timed vaccination response model -- AS03 CITE-seq cohort](#fig3.1)  
9.  [Fig 3. & FigS3 B cell AS03 phenotype analysis](#fig3.2)  
10. [Fig 3. & FigS3 mixed effects timed vaccination response model -- AS03 Validatiton cohort](#fig3.3)  
11. [Fig 3. & FigS3 AS03 specific cell phenotypes figure generation](#fig3.4)  
12. [Fig 4. Define high responder baseline cell phenotypes from multivariate model with enrichment](#fig4.1)  
13. [Fig 4. Correlate expression of baseline high responder phenotypes with plasmablast response](#fig4.2)  
14. [Fig 4.  & FigS4 Construct and visualize high responder multicellular network phenotypes](#fig4.3)  
15. [Fig 4. Early kinetics of baseline states](#fig4.4)  
16. [Fig 4. Analysis of mRNA vaccine data to define induction of high responder phenotypes](#fig4.5)
17. [Fig.5. Define and test AS03 specific cell phenotypes in high responders at baseline](#fig5.1)
18. [Fig.5. Analysis of cell frequency of activated monocyte phenotypes in flow cytometry data](#fig5.2)
19. [Fig.5. Analysis of CyTOF stimulation phenotypes](#fig5.3)
20. [Write output](#output)
21. [Preprocessing to create the starting data](#preprocessing) 

### Instructions for analysis workflow.  <a name="instructions"></a>  

This repository documents all of the code needed to reproduce the analysis results of the manuscript. The data for this analysis in the Zenodo repository below should be added to a data/ directory in the project root folder. The analysis can now run as is without modification--Each R script is self-contained, reading data from the /data folder and writing to figures or results files within each analysis subdirectory relative to the root directory using the R package `here`. Unless otherwise noted, scripts were run with R 4.0.5. Packege versions are listed in the table.  

**Starting raw data for analysis: 10.5281/zenodo.7365959**    


non CITEseq data used in the analysis is saved in the following directories:  
data  
  --vand  
  --CHI_H1N1_data  
  --full_metadata  
  --CHI_H5N1_data  
  --GSE171964  
  --stim  

Annotated CITE-seq data in a Seurat object is saved in the project directory as 
h1h5_annotated_with_meta.rds  
# a version with added sample metadata on each cell is created as below and saved in the first script: 
data
  --h1h5_annotated_with_meta.rds


Many of these scripts require the [scglmmr](https://github.com/MattPM/scglmmr) package associated with this manuscript.    

### Fig1 & Fig S1. sample, cell frequency, and protein distributions  <a name="Fig.1.1"></a>   
*This section is run with R 3.5.1*  
Calculate and visualize distribution of individuals across protein based clusters.  
Biaxial plots of unsupervised cluster dsb normalized protein expression.  
mid_res/sample_and_protein_distributions/1.sample.lineage.proteinbiaxial.r  
```{r, code = readLines(here('mid_res/sample_and_protein_distributions/1.sample.lineage.proteinbiaxial.r'))}
```

Calculate and visualize distribution of immune cell frequency across individuals.  

mid_res/sample_and_protein_distributions/2.cell.frequency.map.r  
```{r, code = readLines(here('mid_res/sample_and_protein_distributions/2.cell.frequency.map.r'))}
```

Clustered protein histogram distribution across cell types

mid_res/histogram_hclust/hclust_histogram_protein.r  
```{r, code = readLines(here('mid_res/histogram_hclust/hclust_histogram_protein.r'))}
```

Aggregated protein expression heatmap across 780 libraries by sample x timepoint  

mid_res/histogram_hclust/hclust_histogram_protein.r  
```{r, code = readLines(here('mid_res/aggregated_protein_libraries/aggregated_ADT_780libraries.r'))} 
```


### transcriptome analysis of manually gated plasmablasts and activated B cells. <a name="fig1.2"></a>   

Analyze transcriptome of manually gated cells using cell type specific gene signatures  
mid_res/pblast_abc_integration/1_pbasc_analysis_gates_modscoresv2.r  
```{r, code = readLines(here('mid_res/pblast_abc_integration/1_pbasc_analysis_gates_modscoresv2.r'))}
```

### Fig 1. multivariate analysis of human and cell type variations. <a name="fig1.3"></a>  

Analyze variance explained by each factor of a multivariate model fit across all cell types. This model includes call type as a variable. 
mid_res/variance_partition/1_variance_partition_allsamples.r  
```{r, code = readLines(here('mid_res/variance_partition/1_variance_partition_allsamples.r'))}
```

Analyze variance explained by each factor of multivariate models fit within each cell protein based type. 
mid_res/variance_partition/2_variance_partition_withincelltype.r  
```{r, code = readLines(here('mid_res/variance_partition/2_variance_partition_withincelltype.r'))}
```

Figure generation from analysis above
mid_res/variance_partition/3_figures_variance_partition_withincelltype.R  
```{r, code = readLines(here('mid_res/variance_partition/3_figures_variance_partition_withincelltype.R'))}
```

Gene set enrichment of genes ranked by variance explained by age in CD8 T cell subsets.    
mid_res/variance_partition/4_age_variancefraction_enrichment.r  
```{r, code = readLines(here('mid_res/variance_partition/4_age_variancefraction_enrichment.r'))}
```

### Fig 2 & Fig S2. mixed effects timed vaccination response model -- unadjuvanted cohort. <a name="fig2.1"></a>  

Mixed effects covariate adjusted model of vaccination effects across donors within each protein based subset -- unadjuvanted cohort. Day 1 and day 7 post vaccination effects estimates. Derive the effect size of vaccination effect to rank genes and run enrichment.

mid_res/1_H1N1_pseudobulk_DE/1_h1_mixed_effect_workflow_V4.r  
```{r, code = readLines(here('mid_res/1_H1N1_pseudobulk_DE/1_h1_mixed_effect_workflow_V4.r'))}
```

Enrichment of curated gene signature pathways based on genes ranked by vaccination effects above.  
mid_res/1_H1N1_pseudobulk_DE/2_rungsea_day1_day7_V4.r  
```{r, code = readLines(here('mid_res/1_H1N1_pseudobulk_DE/2_rungsea_day1_day7_V4.r'))}
```

Generate figures from model results above for day 1 vaccination effects in unadjuvanted cohort. 
Derive dell type specific vaccination genes and shared core interferon signature.  
mid_res/1_H1N1_pseudobulk_DE/3_V4_figures.r  
```{r, code = readLines(here('mid_res/1_H1N1_pseudobulk_DE/3_V4_figures.r'))}
```
Generate figures from model results above for day 7 vaccination effects in unadjuvanted cohort.  
mid_res/1_H1N1_pseudobulk_DE/4_V4_figures_d7.R  
```{r, code = readLines(here('mid_res/1_H1N1_pseudobulk_DE/4_V4_figures_d7.R'))}
```

Visualize core shared across subsets interferon signature defined in script 3.  
mid_res/1_H1N1_pseudobulk_DE/5_shared_core_fin_state.r  
```{r, code = readLines(here('mid_res/1_H1N1_pseudobulk_DE/5_shared_core_fin_state.r'))}
```

### Fig S2 visualization of day 7 post vaccination phenotypes and predictive signature deconvolution <a name="fig2.2"></a>  

Comparison of day 7 signatures predictive of antibody response in microarray and aggregated CITE-seq data  
mid_res/array_bulk_comparison/array_bulk_comparison.R  
```{r, code = readLines(here('mid_res/array_bulk_comparison/array_bulk_comparison.R'))}
```

Single cell deconvolution of predictive signatures across protein based subsets.  
mid_res/d7_predictive_deconvolution/1.d7predictive.score.deconvolution.r  
*This script uses R 3.5.1*  
```{r, code = readLines(here('mid_res/d7_predictive_deconvolution/1.d7predictive.score.deconvolution.r'))}
```

Single cell raw UMI count deconvolution of TNFRSF17 gene.  
mid_res/d7_predictive_deconvolution/2.TNFRSF17.deconvolution.r4.0.5.r   
```{r, code = readLines(here('mid_res/d7_predictive_deconvolution/2.TNFRSF17.deconvolution.r4.0.5.r'))}
```

### Fig 2. bottom up single cell reconstruction of single cell monocyte pseudotime <a name="fig2.3"></a>  
*This section is run with R 3.5.1*  

Construct mRNA-based monocyte single cell latent space with DDRTree. Infer pseudotime using monocle. 

mid_res/monocyte_map/1_monocyte_map.r  
```{r, code = readLines(here('mid_res/monocyte_map/1_monocyte_map.r'))}
```

Integrate pseudotime, surface protein levels and time relative to vaccination with mRNA based pseudotime calculated above to interpret 3 branches.  
mid_res/monocyte_map/2_monocyte_mat_visualization.I.r  
```{r, code = readLines(here('mid_res/monocyte_map/2_monocyte_mat_visualization.I.r'))}
```

Integrate "bottom up" single cell monocye pseudotime reconstruction with "top down" mixed effects vaccine perturbation phenotypes. Within the genes defined by the pseudobulk mixed effects models (the section above) and in the leading edge of curated pathway enrichments, calculate branch dependent differential expression using BEAM. Define categories of genes based on their behavior across single cell pseudotime.
mid_res/monocyte_map/3_monocyte_perturbation_integration.r  
```{r, code = readLines(here('mid_res/monocyte_map/3_monocyte_perturbation_integration.r'))}
```

Pathway enrichment within the integrated pseudotime adn perturbation based gene categories defined above.  
note due to issues the enrichr R package servers, the html web based enrichr server was used here. This script contains linkes to those results and the genes used as input. This script uses the LeadingEdgeIndexed function from scglmmr requiring R 4.0.5. 
mid_res/monocyte_map/4_genecat_enrichr.r  
```{r, code = readLines(here('mid_res/monocyte_map/4_genecat_enrichr.r'))}
```


### Fig 3. & FigS3 mixed effects timed vaccination response model -- AS03 CITE-seq cohort <a name="fig3.1"></a>

Fit combined mixed model with unadjuvanted and adjuvanted subjects and apply contrast to define difference in 24h fold change post vaccination within protein subsets adjusted for age and sex.    
mid_res/3_H1N1_H5N1_joint_pseudobulk_DE/1_V4_AS03_contrastmodel.r  
```{r, code = readLines(here('mid_res/3_H1N1_H5N1_joint_pseudobulk_DE/1_V4_AS03_contrastmodel.r'))}
```

Gene set enrichment and defining As03 specific cell type specific leading edge phenotypes.  
mid_res/3_H1N1_H5N1_joint_pseudobulk_DE/2_V4_AS03_contrastmodel_enrichment.r  
```{r, code = readLines(here('mid_res/3_H1N1_H5N1_joint_pseudobulk_DE/2_V4_AS03_contrastmodel_enrichment.r'))}
```

Calculate log cpm for gene distribution visualization in next script.    
mid_res/3_H1N1_H5N1_joint_pseudobulk_DE/3_V4_calc_logcpm_tidyaveragedata.r  
```{r, code = readLines(here('mid_res/3_H1N1_H5N1_joint_pseudobulk_DE/3_V4_calc_logcpm_tidyaveragedata.r'))}
```

Visualize gene distributions of AS03 specific perturbation effects across time between unadjuvanted and AS03 adjuvanted cohorts.  
mid_res/3_H1N1_H5N1_joint_pseudobulk_DE/5_V4_AS03model_gene_distributions.r  
```{r, code = readLines(here('mid_res/3_H1N1_H5N1_joint_pseudobulk_DE/5_V4_AS03model_gene_distributions.r'))}
```

### Fig 3. & FigS3 B cell AS03 phenotype analysis <a name="fig3.2"></a>  
Further analysis of B cell phenotypes. Single cell model of apoptosis signature and correlation of apoptosis signature with B CD40 Activation signature.    
```{r, code = readLines(here('mid_res/3_H1N1_H5N1_joint_pseudobulk_DE/6_V4_bsignal.r'))}
``` 

Distribution of B cell apoptosis genes from all fitted mixed model z statistics from fold change contrast within naive B cells.  

```{r, code = readLines(here('mid_res/3_H1N1_H5N1_joint_pseudobulk_DE/7_bgenes.r'))}
```

Analysis of surface plasmon resonance data to correlate strain specific with non strain antibody avidity in ASo3 adjuvanted donors. 
mid_res/ru/RU_binding.r  
```{r, code = readLines(here('mid_res/ru/RU_binding.r'))}
```

### Fig 3. & FigS3 mixed effects timed vaccination response model -- AS03 Validatiton cohort <a name="fig3.3"></a>  
Fit mixed model with unadjuvanted and AS03 adjuvanted subjects, apply contrast to define difference in 24h fold change post vaccination on FACS sorted subsets from validation cohort. 
Format data:  
mid_res/vand/1_V4_vand_d1_adjvsnon_DATAFORMAT.r  
```{r, code = readLines(here('mid_res/vand/1_V4_vand_d1_adjvsnon_DATAFORMAT.r'))}
```

Fit model:  
mid_res/vand/2_V4_vand_d1_adjvsnon_MODEL.r  
```{r, code = readLines(here('mid_res/vand/2_V4_vand_d1_adjvsnon_MODEL.r'))}
```

Test enrichment of CITE-seq derived AS03 specific cell phenotypes in the validation cohort.  
mid_res/vand/3_V4_vand_enrCITE_sgnals.r  
```{r, code = readLines(here('mid_res/vand/3_V4_vand_enrCITE_sgnals.r'))}
```

### Fig 3. & FigS3 AS03 specific cell phenotypes validation comparison figures <a name="fig3.4"></a>
Combined figures for the CITE-seq and validation cohort.  
mid_res/combined_contrast/1_combined_contrast_vand_citeseq.r  
```{r, code = readLines(here('mid_res/combined_contrast/1_combined_contrast_vand_citeseq.r'))}
```

### Fig 4. Define high responder baseline cell phenotypes from multivariate model with enrichment <a name="fig4.1"></a>  
Use model output from 1_h1_mixed_effect_workflow_V4.r to rank genes based moderated t test statistics of high vs low responder effect at baseline adjusted for age sex & batch.  

```{r, code = readLines(here('mid_res/baseline_response/1_baseline_gseaV3.r'))}
```
These results read in a set of shortened module / pathway names for visualization that is in the starting data folder
Curate enrichment results pt 1. 
mid_res/baseline_response/2_curate_gseaV3.r
```{r, code = readLines(here('mid_res/baseline_response/2_curate_gseaV3.r'))}
```
Curate enrichment pt 2  
mid_res/baseline_response/3_gsea.vis.r  
```{r, code = readLines(here('mid_res/baseline_response/3_gsea.vis.r'))}
```

Within each subset calculate log cpm of cell type specific leading edge enrichment phenotypes based on high vs low responder model.  
```{r, code = readLines(here('mid_res/baseline_response/4_baseline_exprs_amz_score.r'))}
```

### Fig 4. Correlate expression of baseline high responder phenotypes with plasmablast response <a name="fig4.2"></a>  

For each subject calculate the day 7 fold change of the predictive antibody response signature (array data). 
mid_res/baseline_response/4b_d7FC_response_sig.r  
```{r, code = readLines(here('mid_res/baseline_response/4b_d7FC_response_sig.r'))}
```

Correlate expression of cell type specific baseline states with the day7 response signature.  
mid_res/baseline_response/4c_d7sigFC_vs_baseline_correlation.r  
```{r, code = readLines(here('mid_res/baseline_response/4c_d7sigFC_vs_baseline_correlation.r'))}
```

### Fig 4.  & FigS4 Construct and visualize high responder multicellular network phenotypes<a name="fig4.3"></a>  
Create baseline cell phenotype correlation network and 'calculate shared latent information' for intracellular correlations.  
mid_res/baseline_response/5_baseline_sli_correlation_network.r  
```{r, code = readLines(here('mid_res/baseline_response/5_baseline_sli_correlation_network.r'))}
```

Visualize results of baseline high responder networks, part 1.  
mid_res/baseline_response/5b_network_construction_and_visualization.r  
```{r, code = readLines(here('mid_res/baseline_response/5b_network_construction_and_visualization.r'))}
```
Visualize results of baseline high responder networks, part 2.  
mid_res/baseline_response/5c_network_correlations.r
```{r, code = readLines(here('mid_res/baseline_response/5c_network_correlations.r'))}
```

### Fig 4. Early kinetics of baseline states <a name="fig4.4"></a>  

Fit single cell mixed model to test day 1 post vaccination kinetics of baseline cell phenotypes defined above.  
mid_res/baseline_response/6_sc.kinetic.singlecellmodel.r  
```{r, code = readLines(here('mid_res/baseline_response/6_sc.kinetic.singlecellmodel.r'))}
```

Visualize results of kinetic analysis above.  
mid_res/baseline_response/7_sc.kinetic.singlecellmodel.figures.r  
```{r, code = readLines(here('mid_res/baseline_response/7_sc.kinetic.singlecellmodel.figures.r'))}
```

### Fig 4. Analysis of mRNA vaccine data to define induction of high responder phenotypes <a name="fig4.5"></a>  

Process data from GSE171964.  
mid_res/mrna/mrna_1_setup.r  
```{r, code = readLines(here('mid_res/mrna/mrna_1_setup.r'))}
```

Renormalize raw ADT data with `dsb::ModelNegativeADTnorm()` and manually gate cells.  

```{r, code = readLines(here('mid_res/mrna/mrna_2_gatemono.r'))}
```

Test baseline high responder cell phenotypes in the same subsets before and after mRNA vaccination with mixed model.  
mid_res/mrna/mrna_3_baseline.sig.test.mono.r  
```{r, code = readLines(here('mid_res/mrna/mrna_3_baseline.sig.test.mono.r'))}
```
Run same test in DCs  
mid_res/mrna/mrna_3_signature.test.mdc.r  
```{r, code = readLines(here('mid_res/mrna/mrna_3_signature.test.mdc.r'))}
```

### Fig.5. Define and test AS03 specific cell phenotypes in high responders at baseline <a name="fig5.1"></a>   

Aggergate log cpm in high and low responders at baseline  
mid_res/nat_adj/1.calc.lcpm.baselineH1_as03_modelgenes.r  
```{r, code = readLines(here('mid_res/nat_adj/1.calc.lcpm.baselineH1_as03_modelgenes.r'))}
```

Validate AS03 specificity of combined signatures in external AS03 cohort.  
mid_res/nat_adj/2.AS03.innatesigs.vand.validationcohort.r  
```{r, code = readLines(here('mid_res/nat_adj/2.AS03.innatesigs.vand.validationcohort.r'))}
```

Test AS03 specific mDC and monocyte phenotypes in high vs low responders at baseline.  
mid_res/nat_adj/3.natural.adjuvant.signatures.r  
```{r, code = readLines(here('mid_res/nat_adj/3.natural.adjuvant.signatures.r'))}
```

### Fig.5. Analysis of cell frequency of activated monocyte phenotypes in flow cytometry data <a name="fig5.2"></a>   
Analyze flow cytometry data for differences at baseline in high and low responders. Only test innate cell subsets to focus on hypothesis generated from analysis of CITE-seq data. Further test the identified activated monocyte phenotype for its longitudinal kinetics and day 1 vs baseline fold change difference in high and low responders.  
mid_res/flow_kinetic/flow_cellfreq_kinetics.r  
```{r, code = readLines(here('mid_res/flow_kinetic/flow_cellfreq_kinetics.r'))}
```

### Fig.5. Analysis of CyTOF stimulation phenotypes <a name="fig5.3"></a>  

Visualize stimulated and unstimulated cells identified by HDStim.  
mid_res/stim/visualize_stim_cells.r  
```{r, code = readLines(here('mid_res/cytof_stim/visualize_stim_cells.r'))}
```

Fit mixed effects model of median phospho protein marker expression in classical monocytes pre vs post stimulation and compare effects in high vs low responders.  
```{r, code = readLines(here('mid_res/cytof_stim/stim_test_ag.r'))}
```  

### Write output <a name="output"></a>  
Write results for supplementary tables.   
```{r, code = readLines(here('data_write/Final_script_wite_table_fsc.r'))}
```

### Preprocessing to create the starting data  <a name="preprocessing"></a>  
  
This section uses R version 3.5.1, [CITE-seq-count](https://hoohm.github.io/CITE-seq-Count/) version 1.4.2, and [demuxlet](https://github.com/statgen/demuxlet).  

**Directory 1: /Flu_CITEseq_preprocess/1_preprocessing/**
Contains the pipeline for read mapping demultiplexing etc. Cells are demultiplexed with a combination of genotype based demultiplexing and hashing antibodies. Genotype data is required to run these steps and we therefore provide scripts and outputs. 
```{r, eval=TRUE}
fl = c(
list.files(here('Flu_CITEseq_preprocess/1_preprocessing/'))
) 
data.frame(fl = sort(fl))
```

The pipeline listed above must be run in order. Individual scripts not shown here for brevity but can be accessed in the directory above. Demuxlet was run using genotype data with options on a single HPC core with 20G ram with the following options:  
```
demuxlet --sam /path_to_bam --tag-group CB --tag-UMI UB --vcf /path_to_vcf --min-snp 5 --field GT --geno-error 0.01 --out out_path/ --alpha 0 --alpha 0.5 --group-list /path_to_barcode_list
```

The singlet calls merged with the hashing antibody barcode singlets are then combined and qcd in the final script which saves the object: `h1_h5_merged_seurat_object_demultiplexed_sng.rds`,  a Seurat version 2 object that is further processed in the next step.

**Directory 2: /Flu_CITEseq_normalize_cluster/**  

Script 1: runs normalization of mRNA using scran and normalization of ADTs with [dsb](https://CRAN.R-project.org/package=dsb). 
Flu_CITEseq_normalize_cluster/1_Flu_CITEseq_normalize/1_dsbnorm_prot_scrannorm_rna.R  

```{r, code = readLines(here('Flu_CITEseq_normalize_cluster/1_Flu_CITEseq_normalize/1_dsbnorm_prot_scrannorm_rna.R'))}
```

Cluster cells using protein distance matrix on dsb normalized values without dimensionality reduction.   

```{r, code = readLines(here('Flu_CITEseq_normalize_cluster/2_Flu_CITEseq_cluster/1_merged_h1h5_adt_clustering.r'))}
```

Annotate cells and use add custom gating for purifying cell populations. Add sample metadata and save output data used as starting object. 

```{r, code = readLines(here('Flu_CITEseq_normalize_cluster/2_Flu_CITEseq_cluster/2_joint_cluster_annotation.r'))}
```

This script writes `h1h5_annotated_with_meta.rds` to the project root directory /data it is identical to the object in the data/ directory downloaded from zenodo. This is used as the starting object for downstream analysis. This object is a Seurat version 2 S4 class object, but it is parsed and analyzed with other packages and custom functions including Seurat version 4 (not version 2).  

**Software package versions**  

*R version 4.0.5*  

```
dsb_1.0.2                ggsci_2.9                sp_1.4-5                 SeuratObject_4.1.0       Seurat_4.0.1            
viridis_0.5.1            viridisLite_0.3.0        scglmmr_0.1.0            variancePartition_1.25.6 BiocParallel_1.24.1     
limma_3.46.0             magrittr_2.0.3           lme4_1.1-26              Matrix_1.4-1             emmeans_1.5.4           
Rcpp_1.0.9               HDStIM_0.1.0             here_1.0.1               forcats_0.5.1            stringr_1.4.0           
dplyr_1.0.4              purrr_0.3.4              readr_1.4.0              tidyr_1.1.2              tibble_3.1.8            
ggplot2_3.3.3            tidyverse_1.3.0         
```

*R version 3.5.3*    

```
scglmmr_0.1.0       ggraph_1.0.2        igraph_1.2.4.1      ggsci_2.9           ggridges_0.5.1      monocle_2.10.1     
DDRTree_0.1.5       irlba_2.3.3         VGAM_1.1-1          Biobase_2.42.0      BiocGenerics_0.28.0 viridis_0.5.1      
viridisLite_0.3.0   here_0.1            Seurat_2.3.4        Matrix_1.2-15       cowplot_0.9.4       magrittr_2.0.1     
forcats_0.4.0       stringr_1.4.0       dplyr_0.8.5         purrr_0.3.3         readr_1.3.1         tidyr_1.0.2        
tibble_2.1.1        ggplot2_3.1.1       tidyverse_1.2.1 
```



